File: OOAD - Game Of Bricks/Ball.java
/************************************************************************************/
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
/************************************************************************************/
class Ball implements CONSTANTS
{
	int x, y;				// co-ordinates of the top-left corner and not the co-ordinates of the center of the ball
	int xi, yi;
	Color ballColor = Color.yellow;
/*==========================================================================*/
	Ball()
	{
		init ();
	}
	Ball (int x, int y, int xi, int yi)
	{
		this.x = x;
		this.y = y;
		this.xi = xi;
		this.yi = yi;
	}
/*==========================================================================*/	
	public void init ()
	{
		x = (RIGHT_EDGE - LEFT_EDGE)/2 - RADIUS;
		y = BOTTOM_EDGE - 2 * RADIUS;
		xi = BALL_INCREMENT;
		yi = -BALL_INCREMENT;
	}
/*==========================================================================*/
	public void drawBall(Graphics g)
    {
        g.setColor(ballColor);
        g.fillOval(x, y, 2 * RADIUS, 2 * RADIUS);
    }
/*==========================================================================*/
	// move the ball
	public void moveBall()
	{
		x += xi;
		y += yi;
	} // moveBall
/*==========================================================================*/
	public void moveBall(int padx, int padLen)
	{
		x = padx + padLen/2 - RADIUS;
	}
/*==========================================================================*/
	public void revBallDir()
	{
		flipxi();
		flipyi();
	}
	public void flipxi()	{ xi = -xi; }
	public void flipyi()	{ yi = -yi; }
/*==========================================================================*/
	public void showDir(Graphics g)
	{
		g.setFont(new Font("ARIAL", 1, DIR_FONT));
		g.setColor(Color.white);
		if (isDirRightUp())
		{
			g.drawString(RIGHT_UP, x+DIR_CONST, y-DIR_CONST);
		}
		if (isDirLeftUp())
		{
			g.drawString(LEFT_UP, x-DIR_CONST+5, y-DIR_CONST);
		}
	}
/*==========================================================================*/
	public void changeDir ()
	{
		flipxi();
	}
	public void eraseDir (Graphics g)
	{
		g.setFont(new Font("ARIAL", 1, DIR_FONT));
		g.setColor(Color.black);
		g.fillRect(LEFT_EDGE+5, y-2*DIR_CONST-5, RIGHT_EDGE-LEFT_EDGE-5, 2*DIR_CONST+3);
	}
/*==========================================================================*/
	public boolean isDirRightDown()
	{
		return (xi > 0 && yi > 0);
	}
	public boolean isDirLeftDown()
	{
		return (xi < 0 && yi > 0);
	}
	public boolean isDirRightUp()
	{
		return (xi > 0 && yi < 0);
	}
	public boolean isDirLeftUp()
	{
		return (xi < 0 && yi < 0);
	}
/*==========================================================================*/
	public boolean isNearPaddle()
	{
		return (y +2*RADIUS >= BOTTOM_EDGE);
	}
	public boolean isHitHorScrEd()
	{
		if (y <= TOP_EDGE+5)
		{
			return true;
		}
		return false;
	}
	public boolean isHitVerScrEd()
	{
		if (x <= LEFT_EDGE+5 || x+2*RADIUS >= RIGHT_EDGE)
		{
			return true;
		}
		return false;
	}
} // Ball
/************************************************************************************/

File: PCCP - Game Of Bricks/src/ball.cc
#include "ball.h"
#include "header.h"

Ball :: Ball()
{
	x = COLS/2;
	y = LINES-PAD_HEIGHT-1;
	xi = 1;
	yi = -1;
	color = BALL_COLOR;
}
int Ball :: getx()
{
	return x;
}
int Ball :: gety()
{
	return y;
}
void Ball :: setx(int ax)
{
	x = ax;
}
void Ball :: sety(int ay)
{
	y = ay ;
}
        
int  Ball :: getxi()
{
	return xi;
}

int Ball :: getyi()
{
	return yi;
}
void Ball :: setxi(int i)
{
	xi = i;
}
		
void Ball :: setyi(int i)
{
	yi = i;
}



void Ball :: moveBall()
{
	eraseBall();
	x += xi;
	y += yi;
	putBall();
}
void Ball :: moveBall (int padx, int padlen)
{
	eraseBall();
	eraseDir();
	x = padx + padlen/2;
	showDir();
	putBall();
}

void Ball :: putBall()
{
	attron (COLOR_PAIR(color));
	mvaddch (y, x, BALL_CHAR);
	refresh();
	attroff (COLOR_PAIR(color));
}	
void Ball :: eraseBall()
{
	mvaddch(y, x, WHITE_SPACE);
	refresh();
}

void Ball :: flipyi()
{
	yi = -yi;
}

void Ball :: flipxi()
{
	xi = -xi;
}

void Ball :: setDftDir()
{
	xi = 1;
	yi = -1;
}
void Ball :: init()
{
	x = COLS/2-1;
	y = LINES-PAD_HEIGHT-1;
	xi = 1;
	yi = -1;
	putBall();
}

void Ball :: revBallDir()
{
	flipyi();
	flipxi();
}

int Ball :: isHitHorScrEd()
{
	return (y == TOP_EDGE);
}

int Ball :: isHitVerScrEd()
{
	return (x==LEFT_EDGE || x==RIGHT_EDGE);
}

int Ball :: isDirRightDown()
{
	return (xi==1 && yi==1);
}

int Ball :: isDirLeftDown()
{
	return (xi==-1 && yi==1);
}

int Ball :: isDirRightUp()
{
	return (xi==1 && yi==-1);
}

int Ball :: isDirLeftUp()
{
	return (xi==-1 && yi==-1);
}

void Ball :: showDir()
{
	if (isDirRightUp())
		mvaddch (LINES-PAD_HEIGHT-2, x+1, RIGHT_UP);
	if (isDirLeftUp())
		mvaddch (LINES-PAD_HEIGHT-2, x-1, LEFT_UP);
	refresh();
}

void Ball :: changeDir()
{
	eraseDir();
	flipxi();
	showDir();
}
void Ball :: eraseDir()
{
	if (isDirRightUp())
		mvaddch (LINES-PAD_HEIGHT-2, x+1, WHITE_SPACE);
	if (isDirLeftUp())
		mvaddch (LINES-PAD_HEIGHT-2, x-1, WHITE_SPACE);
	refresh();
}

int Ball :: isNearPad()
{
	return (y == LINES-PAD_HEIGHT-1);
}


